File: bindings/imgui_bundle/demos_python/demos_immapp/demo_chinese_font.py
"""
# Displaying non-Latin text (Chinese, Japanese, Korean, Hebrew, Arabic, ...)

The default bundled fonts (DroidSans, Roboto) only contain Latin / Greek / Cyrillic glyphs.
Any CJK / Hebrew / Arabic text therefore shows up as `???`.

> **Solution:** load a font that contains the glyphs you need, and make it the default font.
>
> This demo loads a CJK font *first* (so it becomes the default font), then merges FontAwesome icons.

## Where to get a font

This demo does not ship a CJK font (a full one is ~10 MB). Download one (TTF or OTF):

* [Noto Sans SC (Simplified Chinese)](https://fonts.google.com/noto/specimen/Noto+Sans+SC)
* [Source Han Sans](https://github.com/adobe-fonts/source-han-sans)

The same approach works for Japanese / Korean / Hebrew / Arabic / Cyrillic: just pick a font that contains those glyphs.

*Tip: CJK fonts have no italic styles. For markdown, merge the regular CJK glyphs into the
italic slots too (CJK will appear upright there).*

*Alternative (zero code): overwrite `assets/fonts/DroidSans.ttf` with your own font, keeping that
exact filename. The default font loader will then pick it up automatically.*
"""
from imgui_bundle import hello_imgui, imgui, imgui_md, immapp, register_demos_assets_folder, ImVec4

# Make the demos_assets/ folder searchable before we look for the font below.
register_demos_assets_folder()

# Edit this to match the font you downloaded (path relative to an assets folder):
CHINESE_FONT = "fonts/NotoSansSC-Regular.otf"
FONT_FILE_PRESENT = hello_imgui.asset_exists(CHINESE_FONT)


def load_fonts() -> None:  # called once by runner_params.callbacks.load_additional_fonts
    if FONT_FILE_PRESENT:
        # Load the CJK font FIRST so it becomes the default font (fonts[0])
        hello_imgui.load_font(CHINESE_FONT, 18.0)
        # Merge FontAwesome icons on top (optional)
        fa = hello_imgui.FontLoadingParams()
        fa.merge_to_last_font = True
        hello_imgui.load_font("fonts/fontawesome-webfont.ttf", 16.0, fa)


def gui() -> None:
    imgui_md.render_unindented(__doc__ or "")
    imgui.separator()
    if FONT_FILE_PRESENT:
        imgui.text(f"Loaded '{CHINESE_FONT}'. Sample text:")
        imgui.text("    Chinese:  你好，世界")
        imgui.text("    Japanese: こんにちは世界")
        imgui.text("    Korean:   안녕하세요 세계")
        imgui.text_disabled("Each line renders only if your font covers that script")
        imgui.text_disabled("(Noto Sans SC covers Chinese; use e.g. Noto Sans KR for Korean)")
    else:
        imgui.text_colored(ImVec4(1.0, 0.4, 0.4, 1.0), f"Font not found: {CHINESE_FONT}")
        imgui.text("Place the downloaded font into one of these folders:")
        for path in hello_imgui.get_assets_search_paths():
            imgui.bullet_text(f"{path}")


def main() -> None:
    runner_params = hello_imgui.RunnerParams()
    runner_params.app_window_params.window_title = "Chinese / non-Latin fonts demo"
    runner_params.app_window_params.window_geometry.size = (800, 600)
    runner_params.callbacks.load_additional_fonts = load_fonts
    runner_params.callbacks.show_gui = gui
    immapp.run(runner_params, immapp.AddOnsParams(with_markdown=True))


if __name__ == "__main__":
    main()

File: bindings/imgui_bundle/demos_cpp/demos_immapp/demo_chinese_font.cpp
// Demo: displaying non-Latin text (Chinese, Japanese, Korean, Hebrew, Arabic, ...)
//
// The explanation is rendered in-app as markdown (with clickable download links).
// In short: load a font that contains the glyphs you need, make it the default font,
// and let the on-demand glyph rasterization (ImGui 1.92) do the rest.

#include "hello_imgui/hello_imgui.h"
#include "immapp/immapp.h"
#include "imgui_md_wrapper/imgui_md_wrapper.h"
#include "demo_utils/api_demos.h"
#include "imgui.h"
#include <string>

// Poor man's fix for C++ late arrival in the unicode party:
//    - C++17: u8"my string" is of type const char*
//    - C++20: u8"my string" is of type const char8_t*
// However, ImGui text functions expect const char*.
#ifdef __cpp_char8_t
#define U8_TO_CHAR(x) reinterpret_cast<const char*>(x)
#else
#define U8_TO_CHAR(x) x
#endif

// Edit this to match the font you downloaded (path relative to an assets folder):
static const std::string CHINESE_FONT = "fonts/NotoSansSC-Regular.otf";

// Documentation shown in the app, as markdown (the download links are clickable).
static const std::string DOC = R"(
# Displaying non-Latin text (Chinese, Japanese, Korean, Hebrew, Arabic, ...)

The default bundled fonts (DroidSans, Roboto) only contain Latin / Greek / Cyrillic glyphs.
Any CJK / Hebrew / Arabic text therefore shows up as `???`.

> **Solution:** load a font that contains the glyphs you need, and make it the default font.
>
> This demo loads a CJK font *first* (so it becomes the default font), then merges FontAwesome icons.

## Where to get a font

This demo does not ship a CJK font (a full one is ~10 MB). Download one (TTF or OTF):

* [Noto Sans SC (Simplified Chinese)](https://fonts.google.com/noto/specimen/Noto+Sans+SC)
* [Source Han Sans](https://github.com/adobe-fonts/source-han-sans)

The same approach works for Japanese / Korean / Hebrew / Arabic / Cyrillic: just pick a font
that contains those glyphs.

*Tip: CJK fonts have no italic styles. For markdown, merge the regular CJK glyphs into the
italic slots too (CJK will appear upright there).*

*Alternative (zero code): overwrite `assets/fonts/DroidSans.ttf` with your own font, keeping that
exact filename. The default font loader will then pick it up automatically.*
)";


// Called once by runnerParams.callbacks.LoadAdditionalFonts
void LoadFonts(bool fontFilePresent)
{
    if (fontFilePresent)
    {
        // Load the CJK font FIRST so it becomes the default font (fonts[0])
        HelloImGui::LoadFont(CHINESE_FONT, 18.f);
        // Merge FontAwesome icons on top (optional)
        HelloImGui::FontLoadingParams fa;
        fa.mergeToLastFont = true;
        HelloImGui::LoadFont("fonts/fontawesome-webfont.ttf", 16.f, fa);
    }
}


void Gui(bool fontFilePresent)
{
    ImGuiMd::RenderUnindented(DOC);
    ImGui::Separator();
    if (fontFilePresent)
    {
        ImGui::Text("Loaded '%s'. Sample text:", CHINESE_FONT.c_str());
        ImGui::TextUnformatted(U8_TO_CHAR(u8"    Chinese:  你好，世界"));
        ImGui::TextUnformatted(U8_TO_CHAR(u8"    Japanese: こんにちは世界"));
        ImGui::TextUnformatted(U8_TO_CHAR(u8"    Korean:   안녕하세요 세계"));
        ImGui::TextDisabled("Each line renders only if your font covers that script");
        ImGui::TextDisabled("(Noto Sans SC covers Chinese; use e.g. Noto Sans KR for Korean)");
    }
    else
    {
        ImGui::TextColored(ImVec4(1.f, 0.4f, 0.4f, 1.f), "Font not found: %s", CHINESE_FONT.c_str());
        ImGui::TextUnformatted("Place the downloaded font into one of these folders:");
        for (const std::string& path : HelloImGui::GetAssetsSearchPaths())
            ImGui::BulletText("%s", path.c_str());
    }
}


int main(int, char**)
{
    ChdirBesideAssetsFolder();  // makes the demos_assets/ folder searchable
    bool fontFilePresent = HelloImGui::AssetExists(CHINESE_FONT);

    HelloImGui::RunnerParams runnerParams;
    runnerParams.appWindowParams.windowTitle = "Chinese / non-Latin fonts demo";
    runnerParams.appWindowParams.windowGeometry.size = {800, 600};
    runnerParams.callbacks.LoadAdditionalFonts = [fontFilePresent]() { LoadFonts(fontFilePresent); };
    runnerParams.callbacks.ShowGui = [fontFilePresent]() { Gui(fontFilePresent); };

    ImmApp::AddOnsParams addons;
    addons.withMarkdown = true;
    ImmApp::Run(runnerParams, addons);

    return 0;
}
